CRM/MembershipExtras/Service/MembershipPeriodType/FixedPeriodTypeCalculator.php

```diff
@@ -57,15 +65,23 @@ public function calculate() {
       $annualProRataCalculation = $settings[SettingField::ANNUAL_PRORATA_CALCULATION_ELEMENT];
       $membershipAmount = $membershipType->minimum_fee;
       $taxAmount = $this->instalmentTaxAmountCalculator->calculateByMembershipType($membershipType, $membershipAmount);
+      $membershipDetails = CRM_Member_BAO_MembershipType::getMembershipType($membershipType->id);
       if ($annualProRataCalculation == self::BY_MONTHS) {
-        $diff = $membershipTypeDurationCalculator->calculateMonthsBasedOnDates($this->startDate);
         $duration = self::TWELVE_MONTHS;
+        $diff = $membershipTypeDurationCalculator->calculateMonthsBasedOnDates($this->startDate, $this->endDate, $this->joinDate);
+        if ($this->isDurationWithInOneYearPeriod($duration, $diff)) {
+          $this->recalCalcuateEndDate();
+        }
+        $diff = $membershipTypeDurationCalculator->calculateMonthsBasedOnDates($this->startDate, $this->endDate, $this->joinDate);
       }
       else {
         $duration  = $membershipTypeDurationCalculator->calculateOriginalInDays();
-        $diff = $membershipTypeDurationCalculator->calculateDaysBasedOnDates($this->startDate);
+        $diff = $membershipTypeDurationCalculator->calculateDaysBasedOnDates($this->startDate, $this->endDate, $this->joinDate);
+        if ($this->isDurationWithInOneYearPeriod($duration, $diff)) {
+          $this->recalCalcuateEndDate();
```

Comment:
If the end date was more by 3 years. Should we handle that case in `recalCalcuateEndDate`?

Reply:
CiviCRM core actually calculates end date for you only 2 years, so your start date is 2018, you will end date will be 2020 so 2020 - 1 is 2019, which should cover this scenario. The scenario is a unique for testing purpose renewal only.